include/frozen/algorithm.h

```diff
@@ -171,19 +171,22 @@ template <std::size_t size> class boyer_moore_searcher {
     if (size == 0)
       return { first, first + size };
 
+    if (size > last - first)
+      return { last, last};  
+
     ForwardIterator iter = first + size - 1;
     while (iter < last) {
       std::ptrdiff_t j = size - 1;
       while (j > 0 && (*iter == needle_[j])) {
         --iter;
         --j;
       }
-      if (*iter == needle_[0])
+      if (*iter == needle_[0] && (iter <= last - size))
```

Comment:
I *think* this extra check (and the call to `std::min` below) are not needed thanks to the extra check in the function prologue.

Reply:
Unfortunately it seems like they are, I believe because if some part of the needle is found but the needle is then longer than the rest of the haystack from that index, it goes past the end.

Reply:
Sorry for the delay. What about this approach?

```
iff --git a/include/frozen/algorithm.h b/include/frozen/algorithm.h
index a543eb3..7872daa 100644
--- a/include/frozen/algorithm.h
+++ b/include/frozen/algorithm.h
@@ -166,12 +166,15 @@ public:
       suffix_table_{build_suffix_table(needle)},
       needle_(needle) {}
 
-  template <class ForwardIterator>
-  constexpr std::pair<ForwardIterator, ForwardIterator> operator()(ForwardIterator first, ForwardIterator last) const {
+  template <class RandomAccessIterator>
+  constexpr std::pair<RandomAccessIterator, RandomAccessIterator> operator()(RandomAccessIterator first, RandomAccessIterator last) const {
     if (size == 0)
-      return { first, first + size };
+      return { first, first };
 
-    ForwardIterator iter = first + size - 1;
+    if (size > (size_t)(last - first))
+      return { last, last };
+
+    RandomAccessIterator iter = first + size - 1;
     while (iter < last) {
       std::ptrdiff_t j = size - 1;
       while (j > 0 && (*iter == needle_[j])) {
@@ -181,9 +184,12 @@ public:
       if (*iter == needle_[0])
         return { iter, iter + size};
 
-      iter += std::max(skip_table_[*iter], suffix_table_[j]);
+      std::ptrdiff_t jump = std::max(skip_table_[*iter], suffix_table_[j]);
+      if (jump > last - iter)
+        return { last, last };
+      iter += jump;
     }
-    return { last, last + size};
+    return { last, last };
   }
 };
```

Reply:
That passes all the tests except the one where a needle is found at the end of the string (find "n*" in "nmnn"), at which point it returns past-the-end, in this sequence
`      if (*iter == needle_[0])
        return { iter, iter + size};`

Reply:
This seems to do the trick

```
diff --git a/include/frozen/algorithm.h b/include/frozen/algorithm.h
index a543eb3..13a960f 100644
--- a/include/frozen/algorithm.h
+++ b/include/frozen/algorithm.h
@@ -166,24 +166,30 @@ public:
       suffix_table_{build_suffix_table(needle)},
       needle_(needle) {}
 
-  template <class ForwardIterator>
-  constexpr std::pair<ForwardIterator, ForwardIterator> operator()(ForwardIterator first, ForwardIterator last) const {
+  template <class RandomAccessIterator>
+  constexpr std::pair<RandomAccessIterator, RandomAccessIterator> operator()(RandomAccessIterator first, RandomAccessIterator last) const {
     if (size == 0)
-      return { first, first + size };
+      return { first, first };
 
-    ForwardIterator iter = first + size - 1;
-    while (iter < last) {
+    if (size > (size_t)(last - first))
+      return { last, last };
+
+    RandomAccessIterator iter = first + size - 1;
+    while (1) {
       std::ptrdiff_t j = size - 1;
       while (j > 0 && (*iter == needle_[j])) {
         --iter;
         --j;
       }
-      if (*iter == needle_[0])
+      if (j == 0 && *iter == needle_[0])
         return { iter, iter + size};
 
-      iter += std::max(skip_table_[*iter], suffix_table_[j]);
+      std::ptrdiff_t jump = std::max(skip_table_[*iter], suffix_table_[j]);
+      if (jump >= last - iter)
+        return { last, last };
+      iter += jump;
     }
-    return { last, last + size};
+    return { last, last };
   }
 };
```

Reply:
Note that there was a bug in my original implementation. I'm very thankful for your effort o/

Reply:
That does seem fine to me! I've pushed this version here, as well as removed the last return since with the modified while(true) loop it could never be reached.